src/Device/Driver/FreeVario.cpp

```diff
@@ -0,0 +1,291 @@
+#include "Device/Driver/FreeVario.hpp"
+#include "Device/Driver.hpp"
+#include "Device/Util/NMEAWriter.hpp"
+#include "Message.hpp"
+#include "NMEA/Checksum.hpp"
+#include "NMEA/Info.hpp"
+#include "NMEA/Derived.hpp"
+#include "NMEA/MoreData.hpp"
+#include "NMEA/InputLine.hpp"
+#include "Units/System.hpp"
+#include "Operation/Operation.hpp"
+#include "LogFile.hpp"
+#include "Interface.hpp"
+#include "CalculationThread.hpp"
+#include "Protection.hpp"
+
+/*
+ * Commands via NMEA from FreeVario Device to XCSoar:
+ *
+ * $PFV,B,S,<double value> - Set Bugs to percent by capacity 0 - 100
+ *
+ * $PFV,M,S,<double value> - Set MC External to defined value
+ * $PFV,M,U                - Set MC External up by 0.1
+ * $PFV,M,D                - Set MC External down by 0.1
+ * $PFV,M,A                - Set MC to Auto [NOT WORKING !]
+ *
+ * $PFV,F,C                - Set Vario FlightMode to Circle
+ * $PFV,F,S                - Set Vario FlightMode to SpeedToFly
+ *
+ * $PFV,Q,S,<double value> - Set QNH to defined value
+ *
+ */
+
+class FreeVarioDevice : public AbstractDevice {
+    Port &port;
+
+public:
+  FreeVarioDevice(Port &_port):port(_port) {}
+
+  bool ParseNMEA(const char *line,NMEAInfo &info) override;
+  static bool PFVParser(NMEAInputLine &line, NMEAInfo &info,Port &port);
+  bool PutMacCready(double mc, OperationEnvironment &env) override;
+  bool PutBugs(double bugs, OperationEnvironment &env) override;
+  bool PutQNH(const AtmosphericPressure &pres,OperationEnvironment &env) override;
+  void OnCalculatedUpdate(const MoreData &basic, const DerivedInfo &calculated) override;
+
+};
+
+
+bool
+FreeVarioDevice::PFVParser(NMEAInputLine &line, NMEAInfo &info, Port &port)
+{
+
+  char buffer[80];
+  NullOperationEnvironment env;
+
+  while (!line.IsEmpty()) {
+
+
+    char type = line.ReadOneChar();
+    char subCommand = line.ReadOneChar();
+
+    if (type == '\0')break;
+    if (subCommand == '\0')break;
+
+       switch (type) {
+             case 'M': {
+               if (subCommand == 'S'){
+
+                   double mcIn;
+                   bool mcReadOk = line.ReadChecked(mcIn);
+
+                   if (subCommand == 'S' && mcReadOk){
+                     info.settings.ProvideMacCready(mcIn,info.clock);
+                   }
+                   else {
+                     sprintf(buffer,"PFV,ERR,%s","Error MC IN value");
+                     PortWriteNMEA(port, buffer, env);
+                   }
+               }
+
+               else if (subCommand == 'U'){
+
+                   double mc = info.settings.mac_cready;
+                   double newmc = mc + 0.1;
```

Comment:
Wait what, the FreeVario sends a command to XCSoar to increase the MacCready value? So it doesn't have its own MacCready setting?